// @backstage/plugin-tech-radar/dist/index.esm.js
import { a as RadarPage } from './esm/RadarPage-B5v6w9DF.esm.js';
export { R as TechRadarComponent } from './esm/RadarPage-B5v6w9DF.esm.js';
export { M as MovedState, T as TechRadarPage, t as plugin, a as techRadarApiRef, t as techRadarPlugin } from './esm/plugin-Bl__HWS2.esm.js';
import '@backstage/core-components';
import '@material-ui/core/Grid';
import '@material-ui/core/Input';
import '@material-ui/core/styles';
import '@material-ui/core/Typography';
import 'react';
import '@backstage/core-plugin-api';
import 'react-use/esm/useAsync';
import '@material-ui/core/Dialog';
import '@material-ui/core/DialogTitle';
import '@material-ui/core/Button';
import '@material-ui/core/DialogActions';
import '@material-ui/core/DialogContent';
import '@material-ui/icons/Link';
import '@material-ui/core/Table';
import '@material-ui/core/TableBody';
import '@material-ui/core/TableCell';
import '@material-ui/core/TableContainer';
import '@material-ui/core/TableHead';
import '@material-ui/core/TableRow';
import '@material-ui/core/Paper';
import '@material-ui/icons/ArrowUpward';
import '@material-ui/icons/ArrowDownward';
import '@material-ui/icons/Adjust';
import '@material-ui/core/Typography/Typography';
import 'color';
import 'd3-force';

const Router = RadarPage;

export { RadarPage, Router };
