// @backstage/plugin-scaffolder-backend/dist/alpha.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var alpha = require('@backstage/plugin-scaffolder-common/alpha');
var pluginPermissionNode = require('@backstage/plugin-permission-node');
var router = require('./cjs/router-CZyu6J68.cjs.js');
var backendPluginApi = require('@backstage/backend-plugin-api');
var backendCommon = require('@backstage/backend-common');
var integration = require('@backstage/integration');
var alpha$2 = require('@backstage/plugin-catalog-node/alpha');
var alpha$1 = require('@backstage/plugin-scaffolder-node/alpha');
require('@backstage/plugin-scaffolder-backend-module-github');
require('@backstage/plugin-scaffolder-backend-module-azure');
require('@backstage/plugin-scaffolder-backend-module-bitbucket');
require('@backstage/plugin-scaffolder-backend-module-bitbucket-cloud');
require('@backstage/plugin-scaffolder-backend-module-bitbucket-server');
require('@backstage/plugin-scaffolder-backend-module-gerrit');
require('@backstage/plugin-scaffolder-backend-module-gitlab');
require('@backstage/plugin-scaffolder-backend-module-gitea');
require('@backstage/errors');
require('zen-observable');
require('@backstage/config');
require('lodash');
require('p-queue');
require('@backstage/catalog-model');
require('@backstage/plugin-scaffolder-common');
require('express');
require('express-promise-router');
require('jsonschema');
require('zod');
require('@backstage/plugin-scaffolder-node');
require('yaml');
require('fs-extra');
require('path');
require('luxon');
require('globby');
require('isbinaryfile');
require('isolated-vm');
require('lodash/get');
require('uuid');
require('winston');
require('nunjucks');
require('stream');
require('prom-client');
require('@backstage/plugin-permission-common');
require('url');
require('os');

const templateConditionExports = pluginPermissionNode.createConditionExports({
  pluginId: "scaffolder",
  resourceType: alpha.RESOURCE_TYPE_SCAFFOLDER_TEMPLATE,
  rules: router.scaffolderTemplateRules
});
const actionsConditionExports = pluginPermissionNode.createConditionExports({
  pluginId: "scaffolder",
  resourceType: alpha.RESOURCE_TYPE_SCAFFOLDER_ACTION,
  rules: router.scaffolderActionRules
});
const createScaffolderTemplateConditionalDecision = templateConditionExports.createConditionalDecision;
const scaffolderTemplateConditions = templateConditionExports.conditions;
const createScaffolderActionConditionalDecision = actionsConditionExports.createConditionalDecision;
const scaffolderActionConditions = actionsConditionExports.conditions;

const scaffolderPlugin = backendPluginApi.createBackendPlugin({
  pluginId: "scaffolder",
  register(env) {
    const addedActions = new Array();
    env.registerExtensionPoint(alpha$1.scaffolderActionsExtensionPoint, {
      addActions(...newActions) {
        addedActions.push(...newActions);
      }
    });
    let taskBroker;
    env.registerExtensionPoint(alpha$1.scaffolderTaskBrokerExtensionPoint, {
      setTaskBroker(newTaskBroker) {
        if (taskBroker) {
          throw new Error("Task broker may only be set once");
        }
        taskBroker = newTaskBroker;
      }
    });
    const additionalTemplateFilters = {};
    const additionalTemplateGlobals = {};
    env.registerExtensionPoint(alpha$1.scaffolderTemplatingExtensionPoint, {
      addTemplateFilters(newFilters) {
        Object.assign(additionalTemplateFilters, newFilters);
      },
      addTemplateGlobals(newGlobals) {
        Object.assign(additionalTemplateGlobals, newGlobals);
      }
    });
    env.registerInit({
      deps: {
        logger: backendPluginApi.coreServices.logger,
        config: backendPluginApi.coreServices.rootConfig,
        lifecycle: backendPluginApi.coreServices.rootLifecycle,
        reader: backendPluginApi.coreServices.urlReader,
        permissions: backendPluginApi.coreServices.permissions,
        database: backendPluginApi.coreServices.database,
        auth: backendPluginApi.coreServices.auth,
        discovery: backendPluginApi.coreServices.discovery,
        httpRouter: backendPluginApi.coreServices.httpRouter,
        httpAuth: backendPluginApi.coreServices.httpAuth,
        catalogClient: alpha$2.catalogServiceRef
      },
      async init({
        logger,
        config,
        lifecycle,
        reader,
        database,
        auth,
        discovery,
        httpRouter,
        httpAuth,
        catalogClient,
        permissions
      }) {
        const log = backendCommon.loggerToWinstonLogger(logger);
        const integrations = integration.ScmIntegrations.fromConfig(config);
        const actions = [
          // actions provided from other modules
          ...addedActions,
          // built-in actions for the scaffolder
          router.createFetchPlainAction({
            reader,
            integrations
          }),
          router.createFetchPlainFileAction({
            reader,
            integrations
          }),
          router.createFetchTemplateAction({
            integrations,
            reader,
            additionalTemplateFilters,
            additionalTemplateGlobals
          }),
          router.createDebugLogAction(),
          router.createWaitAction(),
          // todo(blam): maybe these should be a -catalog module?
          router.createCatalogRegisterAction({ catalogClient, integrations, auth }),
          router.createFetchCatalogEntityAction({ catalogClient, auth }),
          router.createCatalogWriteAction(),
          router.createFilesystemDeleteAction(),
          router.createFilesystemRenameAction()
        ];
        const actionIds = actions.map((action) => action.id).join(", ");
        log.info(
          `Starting scaffolder with the following actions enabled ${actionIds}`
        );
        const router$1 = await router.createRouter({
          logger: log,
          config,
          database,
          catalogClient,
          reader,
          lifecycle,
          actions,
          taskBroker,
          additionalTemplateFilters,
          additionalTemplateGlobals,
          auth,
          httpAuth,
          discovery,
          permissions
        });
        httpRouter.use(router$1);
      }
    });
  }
});

exports.createScaffolderActionConditionalDecision = createScaffolderActionConditionalDecision;
exports.createScaffolderTemplateConditionalDecision = createScaffolderTemplateConditionalDecision;
exports.default = scaffolderPlugin;
exports.scaffolderActionConditions = scaffolderActionConditions;
exports.scaffolderTemplateConditions = scaffolderTemplateConditions;
