// @backstage/plugin-auth-react/dist/index.esm.js
import React, { useMemo, useCallback, useEffect } from 'react';
import { useApi, identityApiRef, fetchApiRef, discoveryApiRef, useApp } from '@backstage/core-plugin-api';
import { useAsync, useMountEffect } from '@react-hookz/web';
import { ErrorPanel } from '@backstage/core-components';
import Button from '@material-ui/core/Button';
import { ResponseError } from '@backstage/errors';

function CookieAuthRedirect() {
  const identityApi = useApi(identityApiRef);
  const [state, actions] = useAsync(async () => {
    const { token } = await identityApi.getCredentials();
    if (!token) {
      throw new Error("Expected Backstage token in sign-in response");
    }
    return token;
  });
  useMountEffect(actions.execute);
  if (state.status === "error" && state.error) {
    return /* @__PURE__ */ React.createElement(React.Fragment, null, "An error occurred: ", state.error.message);
  }
  if (state.status === "success" && state.result) {
    return /* @__PURE__ */ React.createElement(
      "form",
      {
        ref: (form) => form == null ? void 0 : form.submit(),
        action: window.location.href,
        method: "POST"
      },
      /* @__PURE__ */ React.createElement("input", { type: "hidden", name: "type", value: "sign-in" }),
      /* @__PURE__ */ React.createElement("input", { type: "hidden", name: "token", value: state.result }),
      /* @__PURE__ */ React.createElement("input", { type: "submit", value: "Continue" })
    );
  }
  return null;
}

const COOKIE_PATH = "/.backstage/auth/v1/cookie";
const ONE_YEAR_MS = 365 * 24 * 36e5;
function useCookieAuthRefresh(options) {
  const { pluginId } = options != null ? options : {};
  const fetchApi = useApi(fetchApiRef);
  const discoveryApi = useApi(discoveryApiRef);
  const channel = useMemo(() => {
    return "BroadcastChannel" in window ? new BroadcastChannel(`${pluginId}-auth-cookie-expires-at`) : null;
  }, [pluginId]);
  const [state, actions] = useAsync(async () => {
    const apiOrigin = await discoveryApi.getBaseUrl(pluginId);
    const requestUrl = `${apiOrigin}${COOKIE_PATH}`;
    const response = await fetchApi.fetch(`${requestUrl}`, {
      credentials: "include"
    });
    if (!response.ok) {
      if (response.status === 404) {
        return { expiresAt: new Date(Date.now() + ONE_YEAR_MS) };
      }
      throw await ResponseError.fromResponse(response);
    }
    const data = await response.json();
    if (!data.expiresAt) {
      throw new Error("No expiration date found in response");
    }
    return data;
  });
  useMountEffect(actions.execute);
  const retry = useCallback(() => {
    actions.execute();
  }, [actions]);
  const refresh = useCallback(
    (params) => {
      const margin = (1 + 3 * Math.random()) * 6e4;
      const delay = Date.parse(params.expiresAt) - Date.now() - margin;
      const timeout = setTimeout(retry, delay);
      return () => clearTimeout(timeout);
    },
    [retry]
  );
  useEffect(() => {
    if (state.status !== "success" || !state.result) {
      return () => {
      };
    }
    channel == null ? void 0 : channel.postMessage({
      action: "COOKIE_REFRESH_SUCCESS",
      payload: state.result
    });
    let cancel = refresh(state.result);
    const listener = (event) => {
      const { action, payload } = event.data;
      if (action === "COOKIE_REFRESH_SUCCESS") {
        cancel();
        cancel = refresh(payload);
      }
    };
    channel == null ? void 0 : channel.addEventListener("message", listener);
    return () => {
      cancel();
      channel == null ? void 0 : channel.removeEventListener("message", listener);
    };
  }, [state, refresh, channel]);
  if (state.status === "not-executed") {
    return { status: "loading" };
  }
  if (state.status === "loading" && !state.result) {
    return { status: "loading" };
  }
  if (state.status === "loading" && state.error) {
    return { status: "loading" };
  }
  if (state.status === "error" && state.error) {
    return { status: "error", error: state.error, retry };
  }
  return { status: "success", data: state.result };
}

function CookieAuthRefreshProvider(props) {
  const { children, ...options } = props;
  const app = useApp();
  const { Progress } = app.getComponents();
  const result = useCookieAuthRefresh(options);
  if (result.status === "loading") {
    return /* @__PURE__ */ React.createElement(Progress, null);
  }
  if (result.status === "error") {
    return /* @__PURE__ */ React.createElement(ErrorPanel, { error: result.error }, /* @__PURE__ */ React.createElement(Button, { variant: "outlined", onClick: result.retry }, "Retry"));
  }
  return /* @__PURE__ */ React.createElement(React.Fragment, null, children);
}

export { CookieAuthRedirect, CookieAuthRefreshProvider, useCookieAuthRefresh };
