// @backstage/plugin-search/dist/index.esm.js
export { S as SearchModal, a as SearchModalProvider, u as useSearchModal } from './esm/useSearchModal-CaX46YAs.esm.js';
import { r as rootRouteRef } from './esm/SearchType-BDK6qMj4.esm.js';
export { H as HomePageSearchBar, S as Router, b as SearchPage, a as SearchType, c as SidebarSearchModal, s as plugin, s as searchPlugin } from './esm/SearchType-BDK6qMj4.esm.js';
import qs from 'qs';
import React, { useCallback } from 'react';
import { useNavigate } from 'react-router-dom';
import { useRouteRef } from '@backstage/core-plugin-api';
import { useContent, SidebarSearchField } from '@backstage/core-components';
import '@backstage/plugin-search-react';
import '@material-ui/core/Dialog';
import '@material-ui/core/DialogActions';
import '@material-ui/core/DialogContent';
import '@material-ui/core/DialogTitle';
import '@material-ui/core/Divider';
import '@material-ui/core/Grid';
import '@material-ui/core/styles';
import '@material-ui/core/Box';
import '@material-ui/core/Button';
import '@material-ui/core/IconButton';
import '@material-ui/icons/ArrowForward';
import '@material-ui/icons/Close';
import '@backstage/version-bridge';
import 'react-use/esm/useUpdateEffect';
import '@backstage/errors';
import 'react-use/esm/usePrevious';
import '@material-ui/core/Checkbox';
import '@material-ui/core/Chip';
import '@material-ui/core/FormControl';
import '@material-ui/core/InputLabel';
import '@material-ui/core/ListItemText';
import '@material-ui/core/MenuItem';
import '@material-ui/core/Select';
import 'react-use/esm/useEffectOnce';
import '@material-ui/core/Accordion';
import '@material-ui/core/AccordionSummary';
import '@material-ui/core/AccordionDetails';
import '@material-ui/core/Card';
import '@material-ui/core/CardContent';
import '@material-ui/core/CardHeader';
import '@material-ui/core/List';
import '@material-ui/core/ListItem';
import '@material-ui/core/ListItemIcon';
import '@material-ui/icons/ExpandMore';
import '@material-ui/icons/FontDownload';
import 'react-use/esm/useAsync';
import '@material-ui/core/Tab';
import '@material-ui/core/Tabs';

const SidebarSearch = (props) => {
  const searchRoute = useRouteRef(rootRouteRef);
  const { focusContent } = useContent();
  const navigate = useNavigate();
  const handleSearch = useCallback(
    (query) => {
      const queryString = qs.stringify({ query }, { addQueryPrefix: true });
      focusContent();
      navigate(`${searchRoute()}${queryString}`);
    },
    [focusContent, navigate, searchRoute]
  );
  return /* @__PURE__ */ React.createElement(
    SidebarSearchField,
    {
      icon: props.icon,
      onSearch: handleSearch,
      to: "/search"
    }
  );
};

export { SidebarSearch };
